test/merger/test_single_qubit_gates_merger.py

```diff
@@ -0,0 +1,268 @@
+from opensquirrel.ir import Bit, BlochSphereRotation, Float, Qubit
+from opensquirrel.passes.merger import SingleQubitGatesMerger
+from opensquirrel.passes.merger.general_merger import rearrange_barriers
+from test.ir_equality_test_base import modify_circuit_and_check
+
+
+@pytest.fixture(name="merger")
+def merger_fixture() -> SingleQubitGatesMerger:
+    return SingleQubitGatesMerger()
+
+
+def test_single_gate(merger: SingleQubitGatesMerger) -> None:
+    builder1 = CircuitBuilder(1)
+    builder1.Ry(0, Float(1.2345))
+    circuit = builder1.to_circuit()
+
+    builder2 = CircuitBuilder(1)
+    builder2.Ry(0, Float(1.2345))
+    expected_circuit = builder2.to_circuit()
+
+    modify_circuit_and_check(circuit, merger.merge, expected_circuit)
+
+    # Check that when no fusion happens, generator and arguments of gates are preserved.
+    assert isinstance(circuit.ir.statements[0], BlochSphereRotation)
+    assert circuit.ir.statements[0].generator == Ry
+    assert circuit.ir.statements[0].arguments == (Qubit(0), Float(1.2345))
+
+
+def test_two_hadamards(merger: SingleQubitGatesMerger) -> None:
+    builder = CircuitBuilder(4)
+    builder.H(2)
+    builder.H(2)
+    circuit = builder.to_circuit()
+
+    expected_circuit = CircuitBuilder(4).to_circuit()
+
+    modify_circuit_and_check(circuit, merger.merge, expected_circuit)
+
+
+def test_two_hadamards_different_qubits(merger: SingleQubitGatesMerger) -> None:
+    builder1 = CircuitBuilder(4)
+    builder1.H(0)
+    builder1.H(2)
+    circuit = builder1.to_circuit()
+
+    builder2 = CircuitBuilder(4)
+    builder2.H(0)
+    builder2.H(2)
+    expected_circuit = builder2.to_circuit()
+
+    modify_circuit_and_check(circuit, merger.merge, expected_circuit)
+
+
+def test_merge_different_qubits(merger: SingleQubitGatesMerger) -> None:
+    builder1 = CircuitBuilder(4)
+    builder1.Ry(0, Float(math.pi / 2))
+    builder1.Rx(0, Float(math.pi))
+    builder1.Rz(1, Float(1.2345))
+    builder1.Ry(2, Float(1))
+    builder1.Ry(2, Float(3.234))
+    circuit = builder1.to_circuit()
+
+    builder2 = CircuitBuilder(4)
+    builder2.ir.add_gate(BlochSphereRotation(0, axis=(1, 0, 1), angle=math.pi))  # this is Hadamard with 0 phase
+    builder2.Rz(1, Float(1.2345))
+    builder2.Ry(2, Float(4.234))
+    expected_circuit = builder2.to_circuit()
+
+    modify_circuit_and_check(circuit, merger.merge, expected_circuit)
+
+    assert isinstance(circuit.ir.statements[0], BlochSphereRotation)
+    assert circuit.ir.statements[0].is_anonymous  # When fusion happens, the resulting gate is anonymous.
+
+    assert isinstance(circuit.ir.statements[1], BlochSphereRotation)
+    assert circuit.ir.statements[1].generator == Rz  # Otherwise it keeps the same generator and arguments.
+    assert circuit.ir.statements[1].arguments == (Qubit(1), Float(1.2345))
+
+    assert isinstance(circuit.ir.statements[2], BlochSphereRotation)
+    assert circuit.ir.statements[2].is_anonymous
```

Comment:
The same. Is this needed?

@elenbaasc  Why do we expect both that `circuit.ir.statements[2].is_anonymous` and that `expected_circuit.ir.statements[2]` be an `Ry`?

Reply:
Same as before. 

Keep an eye on the qubit indices:

```
Ry(0, math.pi / 2)
Rx(0, math.pi)
Rz(1, 1.2345)
Ry(2, 1)
Ry(2, 3.234)
```

The first 2 on qubit 0 should become an anonymous BSR. The 3rd on qubit 1 should stay an Rz, and the last 2 should become a anonymous BSR as well.

It is checked that the merging of the gates respect the qubit index.

Reply:
OK, I have some input regarding this.

First, I don't think all those asserts are necessary. We build an expected circuit. And it  is enough to check that the modified circuit is equals to that expected circuit.

Second, it indeed happens that, during the process  of merging, the resulting circuit can have a Bloch sphere rotation that is actually an Ry, but that is even though anonymous.

Why? Because this is our implementation for `is_anonymous` (in `ir.py`, class Gate):

```
@property
    def is_anonymous(self) -> bool:
        return self.arguments is None
```

Now, how come do we  get to build a gate which is effectively an `Ry` (or other named gate), but has no arguments? The code that does that is in `general_merger.py`, `compose_bloch_sphere_rotations` function. There, we say:

>     The resulting gate is anonymous except if `a` is the identity and `b` is not anonymous, or vice versa.

And then has the following code:

```
    generator = b.generator if a.is_identity() else a.generator if b.is_identity() else None
    arguments = b.arguments if a.is_identity() else a.arguments if b.is_identity() else None
```

I think we are missing a case here. When we compose a BSR out of 2 named gates with the same generator, e.g., 2 `Ry` gates. I've tried changing the code above to:

```
    if a.is_identity():
        generator = b.generator
        arguments = b.arguments
    elif b.is_identity():
        generator = a.generator
        arguments = a.arguments
    elif a.generator == b.generator:
        generator = a.generator
        arguments = [a.qubit, combined_angle]
    else:
        generator = None
        arguments = None
```

and the last assert from the code above can indeed now be changed to:

```
assert not circuit.ir.statements[2].is_anonymous
```

Reply:
I've created a PR to include a fix for `compose_bloch_sphere_rotations`. You are free to merge that PR into the current PR. But, in any case, after doing that, I would remove all those assertions in the tests after `modify_circuit_and_check`.

Reply:
Another point here is that I don't like very much that implementation/definition of `is_anonymous`... 🤔

Reply:
> Keep an eye on the qubit indices:
> 
> ```
> Ry(0, math.pi / 2)
> Rx(0, math.pi)
> Rz(1, 1.2345)
> Ry(2, 1)
> Ry(2, 3.234)
> ```
> 
> The first 2 on qubit 0 should become an anonymous BSR. The 3rd on qubit 1 should stay an `Rz`, and the last 2 should become a anonymous BSR as well.

I don't think that the last 2 should become an anonymous BSR. It should become an `Ry` with a combined angle of 4.234. And in case it should, then we should build the expected circuit using an anonymous BSR, so that we don't need all the asserts after `modify_circuit_and_check`.